File: leet_code/all_paths_source_to_target.cpp
class Solution {
public:
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        return BFS(graph, 0);
    }
    
    vector<vector<int>> BFS(vector<vector<int>>& graph, int index) {
        vector<vector<int>> paths;
        paths.reserve(graph[index].size());
        if (graph[index].size() == 0) {
            vector<int> base{index};
            paths.push_back(base);
        }
        for (int next : graph[index]) {
            for (vector<int> new_path : this->BFS(graph, next)) {
                new_path.insert(new_path.begin(), index);
                paths.push_back(new_path);
            };
            
        }
        return paths;
    }
    
};

File: leet_code/all_paths_source_to_target.py
class Solution(object):
    def allPathsSourceTarget(self, graph):
        """
        :type graph: List[List[int]]
        :rtype: List[List[int]]
        """
        self.graph = graph
        return self.BFS(0)
    def BFS(self, index):
        paths = []
        if len(self.graph[index]) == 0: return [[index]]
        for next_index in self.graph[index]:
            for subpath in self.BFS(next_index):
                paths.append([index] + subpath)
        return paths


graph = [[1,2], [3], [3], []] 
A = Solution()
print(A.allPathsSourceTarget(graph))